### PR TITLE
ARROW-13345: [C++] Add basic implementation for log to base b

### PR DESCRIPTION
**Pull request by rommelDB**
Changes:
- Added basic implementation for log to base b.
- Added unit tests.

### CODE REVIEW

**Comment by pitrou** [comment]
Is it useful for the case to be given as a `Datum` rather than a function option? Are there use cases where one wants to lookup the log base in a column?

**Comment by rommelDB** [comment]
> Is it useful for the log base to be given as a `Datum` rather than a function option? Are there use cases where one wants to lookup the log base in a column?

@pitrou To be honest, I'm not sure if that such a scenario exists or whether that scenario is frequent. In any case, this draft was my first approach, and I can certainly change it to consider the base as a single value as a function option. Accordingly, I think that the base as a value makes more sense.

**Comment by pitrou** [comment]
cc @ianmcook @jonkeane  for opinions about the log base question.

**Comment by jonkeane** [comment]
That seems pretty uncommon to me. I could contrive a few examples that aren't totally outlandish (say you're scaling/normalizing a value by groups, you might save the log base that you did the normalization with to be able to transform something back into the original values. But that's pretty contrived, and I bet most people would approach that with a lookup table and a for loop, basically.

**Comment by ianmcook** [comment]
I like having the option for the base to be an array. I suspect that in the vast majority of cases users will pass a scalar base, but this implementation is the most flexible way to do it.

**Comment by pitrou** [comment]
> I like having the option for the base to be an array. I suspect that in the vast majority of cases users will pass a scalar base, but this implementation is the most flexible way to do it.

That sounds fine to me.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc [hunk 1]
```diff
@@ -1589,6 +1636,19 @@ const FunctionDoc log1p_checked_doc{
      "-inf or NaN."),
     {"x"}};
 
+const FunctionDoc logb_doc{
+    "Compute log of x to base b of arguments element-wise",
+    ("Values <= -1 return -inf or NaN. Null values return null.\n"
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic.cc:1641]: Values <= 0?

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc [hunk 2]
```diff
@@ -2001,6 +2020,45 @@ TYPED_TEST(TestUnaryArithmeticIntegral, Log) {
...
+  this->AssertBinop(Logb, "[1, 10, null]", "[10, 10, null]",
+                    ArrayFromJSON(float64(), "[0, 1, null]"));
+  this->AssertBinop(Logb, "[1, 2, null]", "[2, 2, null]",
+                    ArrayFromJSON(float64(), "[0, 1, null]"));
+  this->AssertBinop(Logb, "[10, 100, null]", this->MakeScalar(10),
+                    ArrayFromJSON(float64(), "[1, 2, null]"));
+}
+
+TYPED_TEST(TestBinaryArithmeticFloating, Log) {
+  // Integer arguments promoted to double, sanity check here
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:2034]: nit: I don't think this comment applies here

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc [hunk 3]
```diff
@@ -2001,6 +2020,45 @@ TYPED_TEST(TestUnaryArithmeticIntegral, Log) {
...
+  this->AssertBinop(Logb, "[1.0, 10.0, null]", "[10.0, 10.0, null]", "[0.0, 1.0, null]");
+  this->AssertBinop(Logb, "[1.0, 2.0, null]", "[2.0, 2.0, null]", "[0.0, 1.0, null]");
+  this->AssertBinop(Logb, "[10.0, 100.0, 1000.0, null]", this->MakeScalar(10),
+                    "[1.0, 2.0, 3.0, null]");
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:2038]: We should also test what happens on Log of NaN, Inf, and -Inf. We should also check the error cases below here, with both check_overflow = true and check_overflow = false. Or really, it should look similar to the tests for the unary log functions here: https://github.com/apache/arrow/blob/8804fa6c21b9569526e551b92d345820cf63814c/cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc#L1953

**Review by lidavidm — COMMENTED** [review]
Thanks, looks good to me overall! I left some comments on the test cases.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc (hunk 3 above)
Comment (rommelDB) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc:2049]: More unit tests were added.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic.cc (hunk 1 above)
Comment (rommelDB) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic.cc]: Fixed.

File: cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc (hunk 2 above)
Comment (rommelDB) [reply cpp/src/arrow/compute/kernels/scalar_arithmetic_test.cc]: You are right.

**Comment by pitrou** [comment]
@rommelDB Can you try to fix the build errors on Windows (see CI)?

**Review by lidavidm — APPROVED** [review]
LGTM. As Antoine notes, there's a build failure on Windows, you'll likely need a cast somewhere: https://github.com/apache/arrow/pull/10898/checks?check_run_id=3306264947#step:7:901

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Thanks for implementing this!

**Pull request closed without merging**